### DISCUSSION TITLE
Compiling firmware for Pico

### DISCUSSION THREAD

**Discussion by mikek88** [micropython/micropython#10042] [General]
I got a Pico-variant that has 16MB instead of the standard 2MB.  So I duplicated the standard Pico folder and header file, made the changes and compiled a new uf2.  It works fine, but the main.py file does not show up like with the standard Pico firmware.  I'm not sure what is wrong.  My firmware uf2 is the same size as the standard Pico one.

**Comment by scruss** [comment]
If you're not too picky about how the board identifies, you can install the pre-built image for the [Pico LiPo (16MiB)](https://micropython.org/download/PIMORONI_PICOLIPO_16MB/). I've used this on the WeAct 16 MB boards, and it works as expected

**Reply by mikek88** [reply]
I considered that, but I wanted to be able to have the proper firmware.  This 16MB board also has a RGB LED that the Pico LiPo doesn't have.  Although, I should try the Pico LiPo 16MB firmware and see if it loads a main.py file.

**Reply by mikek88** [reply]
I loaded the PicoLipo 16MB firmware and it, too, doesn't create a main.py file.

**Reply by fivdi** [reply]
> I loaded the PicoLipo 16MB firmware and it, too, doesn't create a main.py file.

To the best of my knowledge, flashing MicroPython firmware onto an RP2040 board will not automatically create a file called main.py.

**Reply by mikek88** [reply]
But it does.  Flashing the official Pico firmware onto my board creates the main.py, which runs, and lights the RGB LED and then lights the standard GP25 LED.

**Reply by fivdi** [reply]
I think the main.py you are seeing is from previous usage of MicroPython on the board. If you erase the flash on your Pico (see [
Resetting Flash memory](https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html#resetting-flash-memory)) and then flash MicroPython onto the board again, I'm almost certain that main.py will no longer be there.

**Reply by mikek88** [reply]
Yep, you're right.  I loaded flash_nuke.uf2, reloaded the official firmware and there was no main.py.  I guess the "problem" is that the main.py existed on the 2MB filesystem?...And that the 16MB filesystem couldn't see it?  I guess all is fine, I just no longer have that main.py.

**Reply by fivdi** [reply]
> I guess the "problem" is that the main.py existed on the 2MB filesystem?...And that the 16MB filesystem couldn't see it?

This appears to be have been the case.

**Comment by jimmo** [comment]
If you've previously run 2MiB firmware on this board, this would have created a (2MiB) filesystem, then it's now going to be confused when it goes looking for a 16MiB one with the new firmware.

As long as the filesystem is working, you can just put a new main.py there.

**Reply by mikek88** [reply]
I think the filesystem is file.  I changed the mpconfigboard.h file in /ports/rp2/boards/.  When I checked the filesystem via Thonny and using os.statvfs('/') it reports correct info and not the 2MB Pico numbers.

And, yes, I know I can add a new main.py, but I wanted to properly create firmware that was consistent with all of the others.

**Reply by mikek88** [reply]
@jimmo, Looks like you were right!  I thought the filesystem stats being different was all that mattered.  It looks like the main.py file existed on the 2MB filesystem and going to a 16MB filesystem changed the structure without losing the file.  Though, I've lost the file now, after flash_nuke. lol

**Comment by fivdi** [comment]
If MicroPython is flashed onto a pristine Pico that has never been flashed before, there will be no file called main.py. To the best of my knowledge this is the same for other RP2040 boards too. What exactly are you expecting to occur? Note that I don't use Thonny and I don't know if it automatically creates a main.py on a Pico. If you are having difficulties, there is an application called flash_nuke.uf2 that can be used to erase the content of the flash on RP2040 boards.

**Reply by mikek88** [reply]
I don't think it's Thonny that creates the main.py, because the code run immediately after flashing without having run Thonny.  I want to duplicate this behavior so that when I create my own firmware and flash it I can see the demo code running.

**Reply by fivdi** [reply]
> I want to duplicate this behavior so that when I create my own firmware and flash it I can see the demo code running.

I'm not really sure what the requirement is. What behaviour do you want to duplicate?

**Reply by fivdi** [reply]
Remember, after MicroPython is flashed onto a pristine Pico or other RP2040 board that was never flashed before, there is no main.py in directory `/`. If you are seeing one there, it's because you are flashing MicroPython for at least the second time and the main.py you are seeing was already there from previous usage.

**Reply by fivdi** [reply]
If needed, a tool like mpremote (see [here](https://pypi.org/project/mpremote/) and [here](https://docs.micropython.org/en/latest/reference/mpremote.html)) can be used to copy a file called main.py to your board.

**Comment by mattytrentini** [comment]
For reference, there is a board definition for the [WeAct Studio boards](https://github.com/micropython/micropython/tree/master/ports/rp2/boards/WEACTSTUDIO) (of which 'Wavgat' is one of the popular manufacturers). It's likely (almost certain?) that this definition will suit your board.

We've added _board variants_ to allow 2/4/8/16MB builds for that board definition (see that link above on how to select the variants). 

However, right now, the download manager still needs a change to build board variants for the rp2 port. But you're in luck! The default variant is 16MB so you should be able to use the [WeAct Studio pre-built firmware](https://micropython.org/download/WEACTSTUDIO/). (After following all the other advice here about resetting the filesystem.)
